### PR TITLE
[NVIDIA] Integrate FlashInfer decode kernel (Blackwell) for Qwen3.5

### PR DESCRIPTION
**Pull request by kaixih**
## Summary

Integrate [FlashInfer](https://github.com/flashinfer-ai/flashinfer)'s `gated_delta_rule_decode_pretranspose` kernel as an optional backend (`--gdn-backend flashinfer`) for GDN (Gated Delta Network) layers in Qwen3.5 hybrid linear attention models.

### What changed

- Added `--gdn-backend {triton,cutedsl,flashinfer}` server argument (default: `triton`)
- **FlashInfer decode path**: uses the bf16 pretranspose kernel (`gated_delta_rule_decode_pretranspose`) which operates directly on the K-last state pool via `initial_state`/`initial_state_indices`, eliminating explicit gather/scatter calls on every decode step (https://github.com/flashinfer-ai/flashinfer/pull/2619)
- **FlashInfer prefill path**: uses `chunk_kda` with `initial_state_indices` for gather/scatter into the pool
- Requires FlashInfer ≥ 0.6.4 (Need update!)

### Performance (Qwen3.5-FP8, decode-focused, 256 concurrency, 8xB200)

| Metric | Before | After | Δ |
|---|---|---|---|
| Output throughput (tok/s) | 5638 | 5741 | **+1.8%** |
| TPOT (ms) | 21.51 | 20.96 | **-2.6%** |
| Mean ITL (ms) | 21.58 | 21.04 | **-2.5%** |
| Total throughput (tok/s) | 7047 | 7176 | **+1.8%** |

Per-step profiling: overall decode step 353 µs → 316 µs (−10%), GDN kernel 30 µs → 16 µs (−47%).

The end-to-end gains are modest (~2%) because GDN accounts for a relatively small fraction of the total decode step. The kernel-level improvement is more significant: GDN alone drops by ~47%, contributing ~38% of the overall per-step savings. The remaining improvement comes from other FlashInfer kernel optimizations.

Note: the prefill path still uses KV layout with explicit gather/scatter (the pool->batch trick), which limits the overall e2e gains. A dedicated prefill kernel with native pool support is left for a follow-up PR.

### Kernel microbenchmark — T=1 decode latency (µs)

At batch sizes ≥ 32, the FlashInfer bf16 kernel (this PR) achieves **~2.4–2.8x** speedup over the triton KV reference and **~1.4–1.6x** over the CuteDSL VK kernel from [#17981](https://github.com/sgl-project/sglang/pull/17981). Note, both #17981 and this PR use VK layout.

| B | TRI-KV-BF16 (triton, ref) | SG-VK-BF16 ([#17981](https://github.com/sgl-project/sglang/pull/17981)) | FI-VK-BF16 (this PR) |
|--:|--:|--:|--:|
| 1 | 4.3 | 3.0 | **2.8** |
| 32 | 59.1 | 34.4 | **20.9** |
| 64 | 118.4 | 69.5 | **49.0** |
| 128 | 235.7 | 139.0 | **95.8** |
| 256 | 466.1 | 283.7 | **187.8** |
| 512 | 931.4 | 565.9 | **365.6** |

### Accuracy

| Benchmark | Backend | Accuracy |
|---|---|---|
| GSM8K | flashinfer (this PR) | 0.945 |
| GPQA | flashinfer (this PR) | 0.848, 0.874, 0.884, 0.879, 0.889, 0.838, 0.889, 0.869 (mean: **0.871**) |

### CODE REVIEW

**Comment by kaixih** [comment]
cc. @hlu1

File: python/sglang/srt/layers/attention/linear/kernels/gdn_flashinfer.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import torch
+
+from sglang.srt.layers.attention.linear.kernels.kernel_backend import (
+    LinearAttnKernelBase,
+)
+
+
+class FlashInferGDNKernel(LinearAttnKernelBase):
+    """FlashInfer pretranspose kernel for GDN decode.
```
Comment (yuan-luo) [inline python/sglang/srt/layers/attention/linear/kernels/gdn_flashinfer.py]: This is not necessarily for decode, it can be prefill in the future. Maybe revise the comment.

**Review by ispobock — APPROVED** [review]

**Review by ispobock — APPROVED** [review]
